### PR TITLE
[17.0][MIG] base_tier_validation: Migration to 17.0

### PR DESCRIPTION
**Pull request by celm1990**
superseded by https://github.com/OCA/server-ux/pull/755

### CODE REVIEW

**Comment by celm1990** [comment]
@bosd @moitabenfdz @TonyMasciI @Duranio @pedrobaeza @gurneyalex
The PR is ready for review. I'll squash the commits before merging. For now, it is keep for an easy technical review.

**Comment by bosd** [comment]
@celm1990 Thanks so much for this PR.
Can you please include: https://github.com/OCA/server-ux/commit/127ec31ea7d5edabd9816bd3f60b02505d7397b9

In the other PR, I made  a comment with the propossal to integrate the module `base_tier_validation_waiting` into this base module.
There was no feedback yet...
What is your opinion on this?

Motivation: 
1. The `base_tier_validation_waiting` module might be "too simple" to be an independent module.
2. Mainting it separatly is kinda "costly" it needs some patches to make the unit tests pass. (by either separating the test of the module, or patching all other tier_validation modules)
3. Usability, when making (sequential) tier reviews it makes sense to have an clearoverview on the approval flow.
(Sequence approval is implemented in this base module as well)

**Comment by celm1990** [comment]
> @celm1990 Thanks so much for this PR. Can you please include: [127ec31](https://github.com/OCA/server-ux/commit/127ec31ea7d5edabd9816bd3f60b02505d7397b9)
> 

cherry-pick done

> In the other PR, I made a comment with the propossal to integrate the module `base_tier_validation_waiting` into this base module. There was no feedback yet... What is your opinion on this?
> 

I apologize, but I haven't worked with that module before, and I'm not familiar with its functionality. Additionally, I couldn't find available in versions 16 or 15. Introducing it might require additional migration efforts. I suggest focusing on the current migration for now. In a subsequent pull request, you could propose the integration of that module directly.

**Comment by bosd** [comment]
>I apologize, but I haven't worked with that module before, and I'm not familiar with its functionality.

Sorry, forgot to mention a little bit of background info...
 
_What the module does is quite simple. It adds another validation status `waiting`.
When reaching that status a notification can be sent out.
Use case: In sequential approval flow, you only want to sent  out a notification to the approver when it is 'his' turn._

It's really small change, but because it is adding a status. It's kinda problematic and breaking tests.

Currently I have PR's open for `base_tier_validation_waiting` for version 15 and 16. (for quite some while now)
As per discussion in the pr of V15 https://github.com/OCA/server-ux/pull/638 it become clear that unit tests of that module is kinda problematic.



> I couldn't find available in versions 16 or 15. Introducing it might require additional migration efforts.

**The testing problem**
Could likely be the cause for not yet having this module merged into 15 and 16 branches.
There is no clean, neat and "oca" way to test it. I've discussed this with some dev's at the OXP 2023.
The options:
1. Change functionality when in test context (Dirty but works, current implementation)
2. Create an isolated testing runboat (very dirty hack. and not a real integration test)

The added actual migration effort is a good thing. :nerd_face: 
The best time to refactor is in a migration. As all the code for the depending modules need to be migrated anyway.
If we include it now, we take the pain once, and we know that all depending modules for V17 and onwards are compatible.

> In a subsequent pull request, you could propose the integration of that module directly.

 Doing the refactor in the "middle" of a release will break tests, CI/CD.
(Not only here on the OCA repo's but also on other partners). We will encounter issues when we have to versions of the `base_tier_validation` module. 

Sorry, for thinking out loud a bit :wink: 
The more I think of it, I'm convinced it should be done in the migration to V17.


Preferrably, I would like to have an opinion on this matter from of one the more OCA expeirenced members / heavyweights. With a deeper understanding of these modules.
Is that where the PSC's are for?? 

We could also drag this "issue" along to next versions.
(But as V17 is not used in production yet, we might have a chance here.)

**Comment by bosd** [comment]
While typing my previous comment V16 got merged :sparkles: :tada:  in https://github.com/OCA/server-ux/pull/748

**Comment by celm1990** [comment]
> > I apologize, but I haven't worked with that module before, and I'm not familiar with its functionality.
> 
> Sorry, forgot to mention a little bit of background info...
> 
> _What the module does is quite simple. It adds another validation status `waiting`. When reaching that status a notification can be sent out. Use case: In sequential approval flow, you only want to sent out a notification to the approver when it is 'his' turn._
> 
> It's really small change, but because it is adding a status. It's kinda problematic and breaking tests.
> 
> Currently I have PR's open for `base_tier_validation_waiting` for version 15 and 16. (for quite some while now) As per discussion in the pr of V15 #638 it become clear that unit tests of that module is kinda problematic.
> 
> > I couldn't find available in versions 16 or 15. Introducing it might require additional migration efforts.
> 
> **The testing problem** Could likely be the cause for not yet having this module merged into 15 and 16 branches. There is no clean, neat and "oca" way to test it. I've discussed this with some dev's at the OXP 2023. The options:
> 
> 1. Change functionality when in test context (Dirty but works, current implementation)
> 2. Create an isolated testing runboat (very dirty hack. and not a real integration test)
> 
> The added actual migration effort is a good thing. 🤓 The best time to refactor is in a migration. As all the code for the depending modules need to be migrated anyway. If we include it now, we take the pain once, and we know that all depending modules for V17 and onwards are compatible.
> 
> > In a subsequent pull request, you could propose the integration of that module directly.
> 
> Doing the refactor in the "middle" of a release will break tests, CI/CD. (Not only here on the OCA repo's but also on other partners). We will encounter issues when we have to versions of the `base_tier_validation` module.
> 
> Sorry, for thinking out loud a bit 😉 The more I think of it, I'm convinced it should be done in the migration to V17.
> 
> Preferrably, I would like to have an opinion on this matter from of one the more OCA expeirenced members / heavyweights. With a deeper understanding of these modules. Is that where the PSC's are for??
> 

Thank you for the detailed explanation about the module. It's helpful to have a clear overview of the situation.

I'm not a member of the PSC and don't have the authority to make decisions on merging. @OCA/tools-maintainers  what do you think, it is possible merge `base_tier_validation_waiting` into `base_tier_validation`

File: base_tier_validation/models/res_config_settings.py
Comment (moitabenfdz) [inline base_tier_validation/models/res_config_settings.py:1]: Is the commented code still relevant?

File: base_tier_validation/models/res_users.py
Comment (moitabenfdz) [inline base_tier_validation/models/res_users.py]: Clean line 36 commented code.

File: base_tier_validation/views/res_config_settings_views.xml
Comment (moitabenfdz) [inline base_tier_validation/views/res_config_settings_views.xml]:
```suggestion
                     <block
                        class="row mt16"
                        name="base_tier_validation_option_setting_container"
                        title="Tier Validation"
                    >
                        <setting
                            class="col-12 col-lg-6"
                            id="module_base_tier_validation_formula"
                        >
                            <div>
                                <field name="module_base_tier_validation_formula" />
                                <label for="module_base_tier_validation_formula" />
                                <div class="text-muted">
                                    Option to use python formula to get reviewers and find documents
                                </div>
                            </div>
                        </setting>
                        <setting
                            class="col-12 col-lg-6"
                            id="module_base_tier_validation_forward"
                        >
                            <div>
                                <field name="module_base_tier_validation_forward" />
                                <label for="module_base_tier_validation_forward" />
                                <div class="text-muted">
                                    Option to forward a tier validation to other reviewer, and backward
                                </div>
                            </div>
                        </setting>
                    </block>
```

This changes are proposed in the [migration guidelines](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0)

Also, I suggest deleting the commented code (check if it's no longer needed).

**Review by moitabenfdz — APPROVED** [review]
I think the changes would be good, but it works as it is, so I will approve the PR.

File: base_tier_validation/models/res_config_settings.py
Comment (celm1990) [reply base_tier_validation/models/res_config_settings.py:1]: When the respective modules are migrated, this code must be uncommented. Currently, I have commented out other fields that are related to modules that have not been migrated yet.

**Comment by bosd** [comment]
> @OCA/tools-maintainers what do you think, it is possible merge `base_tier_validation_waiting` into `base_tier_validation`

@dreispt Can you help us out? on this one?

**Comment by celm1990** [comment]
@pedrobaeza @LoisRForgeFlow, could you review this PR please

**Comment by pedrobaeza** [comment]
Sorry, I'm not the more suitable to review this one.

/ocabot migration base_tier_validation

**Review by bosd — CHANGES_REQUESTED** [review]
By the way.. there are still some todo's open for V17

File: base_tier_validation/models/tier_validation.py
Comment (bosd) [inline base_tier_validation/models/tier_validation.py:35]: Sorry.. forgot to provide example of those lines..

**Comment by bosd** [comment]
FYI, I've started working on merging the modules and thus improving sequential message flow.
based it upon this pr. Hope to show something in the next couple of days.

**Comment by celm1990** [comment]
@bizzappdev @StefanRijnhart @dreispt  could you please review this PR 😅

**Comment by bizzappdev** [comment]
@celm1990 You might need to squash all migration commits into one commit.

**Comment by sonhd91** [comment]
LGTM

**Review by sonhd91 — APPROVED** [review]
Need squash UDP commit

**Comment by celm1990** [comment]
> @celm1990 You might need to squash all migration commits into one commit.

Done

**Comment by bosd** [comment]
Please include #724

File: base_tier_validation/models/tier_validation.py
Comment (bosd) [reply base_tier_validation/models/tier_validation.py:35]: @victoralmau Seems like yoiu've added these comments.
Can you clarify / confirm which fiuelds you are talking about..?
is it about `validated` / `rejected`field?

Comment (pedrobaeza) [reply base_tier_validation/models/tier_validation.py:35]: He's on holidays this week. The idea is to have a `validation_status` selection field that unifies the logic for allowing to validate, instead of several boolean fields.

File: base_tier_validation/security/ir.model.access.csv
Comment (nguyenminhchien) [inline base_tier_validation/security/ir.model.access.csv]: I faced this warning
```

2024-01-09 08:42:29,782 12362 WARNING v17c_base_tier_validation odoo.addons.base.models.ir_model: Rule access.tier.review has no group, this is a deprecated feature. Every access-granting rule should specify a group. 
2024-01-09 08:42:29,782 12362 WARNING v17c_base_tier_validation odoo.addons.base.models.ir_model: Rule tier.definition.all has no group, this is a deprecated feature. Every access-granting rule should specify a group. 
2024-01-09 08:42:29,782 12362 WARNING v17c_base_tier_validation odoo.addons.base.models.ir_model: Rule access.comment.wizard has no group, this is a deprecated feature. Every access-granting rule should specify a group. 

```

File: base_tier_validation/static/description/index.html
Comment (nguyenminhchien) [inline base_tier_validation/static/description/index.html:1]: I faced this warning, should remove this line
```

2024-01-09 08:42:30,778 12362 INFO v17c_base_tier_validation odoo.modules.registry: module base_tier_validation_formula: creating or updating database tables 
2024-01-09 08:42:30,872 12362 WARNING v17c_base_tier_validation py.warnings: /home/chien/code/odoo/odoo/17.0/odoo/addons/base/models/ir_module.py:170: DeprecationWarning: XML declarations in HTML module descriptions are deprecated since Odoo 17, base_tier_validation_formula can just have a UTF8 description with not need for a declaration.

```

File: base_tier_validation/__manifest__.py
Comment (nguyenminhchien) [inline base_tier_validation/__manifest__.py]: the path "/base_tier_validation/static/src/xml/**/*" did not resolve to anything.

File: base_tier_validation/models/tier_validation.py
Comment (nguyenminhchien) [inline base_tier_validation/models/tier_validation.py]: try this instead:
`                    lambda r, x=rec: r.definition_id.notify_on_create and r.res_id == x.id`

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]

File: base_tier_validation/static/description/index.html
Comment (bosd) [reply base_tier_validation/static/description/index.html:1]: This one should be fixed when repo update in #798 is merged.

File: base_tier_validation/security/ir.model.access.csv
Comment (celm1990) [reply base_tier_validation/security/ir.model.access.csv]: done

File: base_tier_validation/models/tier_validation.py
Comment (celm1990) [reply base_tier_validation/models/tier_validation.py]: thanks

File: base_tier_validation/static/description/index.html
Comment (celm1990) [reply base_tier_validation/static/description/index.html:1]: This is just a warning, @pedrobaeza do you have any plans to update the global OCA update copier after merging this PR https://github.com/OCA/oca-addons-repo-template/pull/239?

**Comment by celm1990** [comment]
@nguyenminhchien could you please update your review

**Review by nguyenminhchien — APPROVED** [review]

File: base_tier_validation/static/description/index.html
Comment (pedrobaeza) [reply base_tier_validation/static/description/index.html:1]: Oh, I didn't remember that it was fixed at maintainer-tools level. @sbidoul has the token to update with the bot all the branches.

**Comment by bosd** [comment]
Oops, this merge went quick.
The V17 comments where not yet adressed.
I was about to open the merge pr based on this one this evening. 

Well, we can discuss that merge later in that pr :wink:

Anyway big thanks @celm1990 for all your work on this pr!

**Pull request merged**
